### PR TITLE
Optimizations to recordPrevalence and related functions

### PR DESCRIPTION
**Pull request by alexristich**
Here's a first stab at the refactoring of the `recordPrevalence` call path. It turns out most of the code from `heuristicBlockingAccounting` (didn't modify the name here, but we should at some point) was superfluous when operating under the assumption that anything in `snitch_map` from an import was already confirmed to have been tracking (as are any calls from `recordFingerprinting`).

Above is the "Partial" optimization. The "Full" optimization includes the other optimization present in this pull, namely the removal of two calls to `setupHeuristicAction` for `tracker_fqdn` and `tracker_origin` in `heuristicblocking.js` line 204:

```javascript
if(firstParties.indexOf(page_origin) === -1){
  firstParties.push(page_origin);
  snitch_map.setItem(tracker_origin, firstParties);
} 
```

Benchmarks are as follows, with "Runtime" being the time that the Privacy Badger extension runs at 100% CPU. Set size is 600 tracker domains with 8300 first parties:

| Version    | Test        | Runtime   |
| :------------- |:------------- |:----- |
| Current      | Fresh import | 3m30s |
| Opt(Partial) | Fresh import   |    50s |
| Opt(Full) | Fresh import   |    0s |
| Current      | Re-import      |   3m30s |
| Opt(Partial) | Re-import      |  0s |
| Opt(Full) | Re-import   |    0s |

Still, the removal of those two lines from `heuristicblocking.js` causes tests to fail and I'm not sure I entirely understand why. They appear to be responsible for the last major vestige of computational overhead associated with importing, so any help in better explaining why we need these calls would be super helpful!

### CODE REVIEW

File: src/heuristicblocking.js [hunk 1]
```diff
@@ -181,9 +204,6 @@ HeuristicBlocker.prototype = {
     if(firstParties.indexOf(page_origin) === -1){
       firstParties.push(page_origin);
       snitch_map.setItem(tracker_origin, firstParties);
-      // TODO removing this line/messing up parameters should break integration tests
-      this.storage.setupHeuristicAction(tracker_fqdn, constants.ALLOW);
-      this.storage.setupHeuristicAction(tracker_origin, constants.ALLOW);
```
Comment (ghostwords) [inline src/heuristicblocking.js]: Was this removal intentional?

Comment (ghostwords) [inline src/heuristicblocking.js]: Oh yes, sorry, I see now.

Comment (ghostwords) [inline src/heuristicblocking.js]: I think these lines are still needed ... What they do is register the tracker's FQDN (`ssl.google-analytics.com`) and origin (`google-analytics.com`) in `action_map`, so that the user sees them (as unblocked) in the browser popup.

Comment (ghostwords) [inline src/heuristicblocking.js]: This happens the first time we see a tracker (third party resource that had `hasTracking` return true) on a given page origin. Which seems kind of funky: doesn't this only need to happen the very first time we see a potential tracker, not once per page origin?

(Also what if it's a new origin for a known tracker? Then we reset the tracker's status (everywhere) to ALLOW? Well, no, since we should never get this far because of the `if (action != constants.NO_TRACKING && action != constants.ALLOW)` checks in calling functions.)

Seems like we should restore, plus document this logic, and fix or add a note to fix when it runs. My TODO comment is about it not having automated tests.

Comment (ghostwords) [inline src/heuristicblocking.js]: As for tests, QUnit is failing because `action_map.getItem('testsite.com').heuristicAction === "allow"` is no longer truthy, which makes sense.

Comment (ghostwords) [inline src/heuristicblocking.js]: By the way, this logic was introduced in https://github.com/EFForg/privacybadger/pull/819/commits/e225ad60b88d99b62315efd1e54629648e9ef2af as part of [de-ABPification](https://github.com/EFForg/privacybadger/pull/819).

**Comment by alexristich** [comment]
Seems to actually be quite fast in this form. 100% CPU time looks to have been at most a second or two. I'm having trouble running the tests on my local machine again - I'm getting the error `Xvfb did not start`. (@cooperq Do you remember how I managed to fix this last time? I neglected to write it down and am regretting it right about now!)

**Comment by ghostwords** [comment]
To run QUnit tests, just [visit the tests.html page in your browser](https://github.com/EFForg/privacybadger/blob/master/README.md#testing).

**Comment by ghostwords** [comment]
Here's a first pass at a debounced-per-argument function:
```javascript
var debouncedFunc = (function () {
  var funcs = {}; // preserved by the closure
  return function (a) {
    if (!funcs.hasOwnProperty(a)) {
      funcs[a] = _.debounce(function () {
        FUNCTION_BEING_DEBOUNCED(a);
      });
    }
    funcs[a]();
  };
}());
```

**Comment by alexristich** [comment]
@ghostwords Thanks! I can't run any Selenium tests, but I probably should've looked at the README sooner to figure out how to run the QUnit tests :)

Just pushed a change to one of the tests. It was failing because we were manually adding it to `action_map` without a `heuristicAction`. I don't know off the top of my head if there are cases where items are being added to `action_map` without a heuristicAction being set. 

Perhaps another change needs to be made too, though - I'm not convinced the current logic is correct. Specifically, if a domain has `heuristicAction: "notracking"`, it would be passed in to `_recordPrevalence()` but would retain its `notracking` title until it hit the threshold. Does it make sense to also check to see if `heuristicAction === constants.NO_TRACKING` then change it to `constants.ALLOW`? We know that once a domain has reached `_recordPrevalence()` we are already assured that it is tracking so perhaps its a safe bet?

**Comment by alexristich** [comment]
By the way, the test that is failing here seems to be failing on the current master branch as well. I'm not sure what is causing it, but at first glance it doesn't appear that anything in this PR is breaking any tests that are currently passing on master.

File: src/heuristicblocking.js [hunk 2]
```diff
@@ -211,6 +211,13 @@ HeuristicBlocker.prototype = {
       snitch_map.setItem(tracker_origin, firstParties);
     }
 
+    var action_map = this.storage.getBadgerStorageObject('action_map');
+    // If tracker_origin is not yet in action_map, add it with action `constants.ALLOW`
+    if (!action_map.hasItem(tracker_origin)) {
```
Comment (ghostwords) [inline src/heuristicblocking.js]: @cooperq @alexristich Should this be `tracker_fqdn` instead? Otherwise we never initialize FQDNs past the first for an origin ...

Comment (alexristich) [inline src/heuristicblocking.js]: Ah, good point. In fact, `tracker_fqdn` needs to replace `tracker_origin` in a few of the calls.`snitch_map.hasItem()` should be called on the fqdn, as should the `snitch_map.setItem()` call. For this if clause, perhaps there should actually be two separate if clauses, one for `tracker_origin` and another for `tracker_fqdn`. Does that sound about right?

**Comment by ghostwords** [comment]
@cooperq I don't think we should include this in today's release. I would like to give it another round of consideration tomorrow.

File: src/heuristicblocking.js (hunk 2 above)
Comment (ghostwords) [inline src/heuristicblocking.js]: No ... PB records tracking by origin in snitch map: "snitch_map is our collection of potential tracking base_domains. The key is a base domain (ETLD+1) ..."

Comment (ghostwords) [inline src/heuristicblocking.js]: Fixed in e7f396b613a01dc8a9096d5cc088a41deb04a29f. @alexristich Do we even need this check anymore, given that we debounced `_syncStorage`?

Comment (alexristich) [inline src/heuristicblocking.js]: Right, that makes sense. I was rushing out the door and probably should've taken a moment to read the code more carefully!

I suppose we don't need the check anymore for performance reasons; it's just weirdly structured code though so maybe the check helps with readability. I'll let you confirm or deny that, though :)

**Comment by ghostwords** [comment]
OK, I'm done for now, please take a look at my changes. I'll re-review tomorrow with fresh eyes and sign off then hopefully.

**Comment by alexristich** [comment]
On first pass, this definitely looks better - also, good call on adding a link to this discussion for future reference.
